slatedb/src/wal_buffer.rs

```diff
@@ -490,16 +420,83 @@ impl WalBufferManager {
     }
 
     pub async fn close(&self) -> Result<(), SlateDBError> {
-        self.quit_once.write(Ok(()));
-
-        Ok(())
+        self.cancellation_token.cancel();
+        if let Some(background_task) = {
+            let mut inner = self.inner.write();
+            inner.background_task.take()
+        } {
+            background_task
+                .await
+                .expect("failed to join wal buffer task")
+        } else {
+            Ok(())
+        }
     }
 }
 
+#[derive(Debug)]
 struct WalFlushWork {
     result_tx: Option<oneshot::Sender<Result<(), SlateDBError>>>,
 }
 
+struct WalFlushHandler {
+    max_flush_interval: Option<Duration>,
+    wal_buffer: Arc<WalBufferManager>,
+}
+
+#[async_trait]
+impl MessageHandler<WalFlushWork> for WalFlushHandler {
+    fn tickers(&mut self) -> Vec<(Duration, Box<MessageFactory<WalFlushWork>>)> {
+        if let Some(max_flush_interval) = self.max_flush_interval {
+            return vec![(
+                max_flush_interval,
+                Box::new(|| WalFlushWork { result_tx: None }),
+            )];
+        }
+        vec![]
+    }
+
+    async fn handle(&mut self, message: WalFlushWork) -> Result<(), SlateDBError> {
+        let WalFlushWork { result_tx } = message;
+        if let Some(result_tx) = result_tx {
+            let result = self.wal_buffer.do_flush().await;
+            result_tx
+                .send(result.clone())
+                .expect("failed to send flush result");
+            result
+        } else {
+            self.wal_buffer.do_flush().await
+        }
+    }
+
+    async fn cleanup(
+        &mut self,
+        mut messages: BoxStream<'async_trait, WalFlushWork>,
+        result: Result<(), SlateDBError>,
+    ) -> Result<(), SlateDBError> {
+        // drain remaining messages
+        while let Some(WalFlushWork { result_tx }) = messages.next().await {
```

Comment:
This logic is slightly different than the previous implementation, which used `quit_once`. I think the outcome is the same, though.